src/main/java/com/minecolonies/colony/buildings/BuildingFarmer.java

```diff
@@ -4,89 +4,399 @@
+                {
+                    colonyField.setOwner(citizen.getName());
+                }
+                field.setOwner(citizen.getName());
+            }
+        }
+        return new JobFarmer(citizen);
     }
 
+    //we have to update our field from the colony!
     @Override
     public void readFromNBT(NBTTagCompound compound)
     {
         super.readFromNBT(compound);
-
-        NBTTagCompound farmerCompound = compound.getCompoundTag(TAG_FARMER);
-
-        wheat = farmerCompound.getInteger(WHEAT_TAG);
-        potato = farmerCompound.getInteger(POTATO_TAG);
-        carrot = farmerCompound.getInteger(CARROT_TAG);
-        melon = farmerCompound.getInteger(MELON_TAG);
-        pumpkin = farmerCompound.getInteger(PUMPKIN_TAG);
+        final NBTTagList fieldTagList = compound.getTagList(TAG_FIELDS, Constants.NBT.TAG_COMPOUND);
+        for (int i = 0; i < fieldTagList.tagCount(); ++i)
+        {
+            final NBTTagCompound fieldCompound = fieldTagList.getCompoundTagAt(i);
+            final Field f = Field.createFromNBT(getColony(), fieldCompound);
+            if (f != null)
+            {
+                farmerFields.add(f);
+            }
+        }
+        assignManually = compound.getBoolean(TAG_ASSIGN_MANUALLY);
     }
 
     @Override
     public void writeToNBT(NBTTagCompound compound)
     {
         super.writeToNBT(compound);
+        final NBTTagList fieldTagList = new NBTTagList();
+        for (final Field f : farmerFields)
+        {
+            final NBTTagCompound fieldCompound = new NBTTagCompound();
+            f.writeToNBT(fieldCompound);
+            fieldTagList.appendTag(fieldCompound);
+        }
+        compound.setTag(TAG_FIELDS, fieldTagList);
+        compound.setBoolean(TAG_ASSIGN_MANUALLY, assignManually);
+    }
 
-        NBTTagCompound farmerCompound = new NBTTagCompound();
+    /**
+     * Method to serialize data to send it to the view.
+     *
+     * @param buf the used ByteBuffer.
+     */
+    @Override
+    public void serializeToView(ByteBuf buf)
+    {
+        super.serializeToView(buf);
+        buf.writeBoolean(assignManually);
 
-        farmerCompound.setInteger(WHEAT_TAG,wheat);
-        farmerCompound.setInteger(POTATO_TAG,potato);
-        farmerCompound.setInteger(CARROT_TAG,carrot);
-        farmerCompound.setInteger(MELON_TAG,melon);
-        farmerCompound.setInteger(PUMPKIN_TAG, pumpkin);
+        int size = 0;
 
-        compound.setTag(TAG_FARMER, farmerCompound);
+        for(Field field: getColony().getFields().values())
+        {
+            if(field.isTaken())
+            {
+                if(getWorker() == null || field.getOwner().equals(getWorker().getName()))
+                {
+                    size++;
+                }
+            }
+            else
+            {
+                size++;
+            }
+        }
+
+        buf.writeInt(size);
+
+        for(Field field: getColony().getFields().values())
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'field' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)